Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp

```diff
@@ -0,0 +1,1320 @@
+      ++i;
+    }
+  }
+  m_target->checkFunction();
+  size_t i = nSpec;
+  for (auto &propFun : compositePhysProps) {
+    multiDomain->addFunction(propFun);
+    multiDomain->setDomainIndex(i, i);
+    ++i;
+  }
+}
+
+/// Calculate excitations at given temperature
+void CrystalFieldFunction::calcExcitations(
+    int nre, const DoubleFortranVector &en, const ComplexFortranMatrix &wf,
+    double temperature, FunctionValues &values, double intensityScaling) const {
+  IntFortranVector degeneration;
+  DoubleFortranVector eEnergies;
+  DoubleFortranMatrix iEnergies;
+  const double de = getAttribute("ToleranceEnergy").asDouble();
+  const double di = getAttribute("ToleranceIntensity").asDouble();
+  DoubleFortranVector eExcitations;
+  DoubleFortranVector iExcitations;
+  calculateIntensities(nre, en, wf, temperature, de, degeneration, eEnergies,
+                       iEnergies);
+  calculateExcitations(eEnergies, iEnergies, de, di, eExcitations,
+                       iExcitations);
+  const auto nPeaks = eExcitations.size();
+  values.expand(2 * nPeaks);
+  for (size_t i = 0; i < nPeaks; ++i) {
+    values.setCalculated(i, eExcitations.get(i));
+    values.setCalculated(i + nPeaks, iExcitations.get(i) * intensityScaling);
+  }
+}
+
+/// Build a function for a single spectrum.
+API::IFunction_sptr CrystalFieldFunction::buildSpectrum(
+    int nre, const DoubleFortranVector &en, const ComplexFortranMatrix &wf,
+    double temperature, double fwhm, size_t iSpec, bool addBackground,
+    double intensityScaling) const {
+  FunctionValues values;
+  calcExcitations(nre, en, wf, temperature, values, intensityScaling);
+  const auto fwhmVariation = getAttribute("FWHMVariation").asDouble();
+  const auto peakShape = getAttribute("PeakShape").asString();
+  auto bkgdShape = getAttribute("Background").asUnquotedString();
+  const size_t nRequiredPeaks = getAttribute("NPeaks").asInt();
+  const bool fixAllPeaks = getAttribute("FixAllPeaks").asBool();
+
+  auto spectrum = new CompositeFunction;
+
+  if (addBackground && !bkgdShape.empty()) {
+    if (bkgdShape.find("name=") != 0 && bkgdShape.front() != '(') {
+      bkgdShape = "name=" + bkgdShape;
+    }
+    auto background =
+        API::FunctionFactory::Instance().createInitialized(bkgdShape);
+    spectrum->addFunction(background);
+  }
+
+  auto xVec = m_control.getFunction(iSpec)->getAttribute("FWHMX").asVector();
+  auto yVec = m_control.getFunction(iSpec)->getAttribute("FWHMX").asVector();
+  CrystalFieldUtils::buildSpectrumFunction(*spectrum, peakShape, values, xVec,
+                                           yVec, fwhmVariation, fwhm,
+                                           nRequiredPeaks, fixAllPeaks);
+  return IFunction_sptr(spectrum);
+}
+
+API::IFunction_sptr CrystalFieldFunction::buildPhysprop(
+    int nre, const DoubleFortranVector &en, const ComplexFortranMatrix &wf,
+    const ComplexFortranMatrix &ham, const std::string &propName) const {
+
```

Comment:
could make a function containing the below if statements and remove the returns. Then the new function can be called below with a get propFun and then returning it. Then the same function can be used in https://github.com/mantidproject/mantid/pull/20389/files#diff-e713a6a7351e91d3ef66099a149c18dfR936

Reply:
The reason is that this method creates functions for physical properties while the method below updates already created functions.